Allow DateOffset addition with Series

chris-b1 opened the pull request:
Addresses #10699 - allowing relative offsets like `DateOffset(years=1)` to to be added to a `Series` containing datetimes for element-wise addition.

Previously only non-relative offsets that could be mapped to a time_delta (`Tick`, `Day`, etc) worked.

Some performance numbers are below - `Tick` based offsets should be unchanged, a subset of offsets (`DateOffset`, `YearEnd/Begin`, `MonthEnd/Begin`, `QuarterEnd/Begin`, `Week`, `BusinessDay`) will be sped up significantly; and other offsets will still fallback to a slow path, but now will raise a performance warning.
### Master

```
In [1]: idx = pd.date_range(start='20140101', freq='T', periods=100000)

In [2]: %timeit idx + pd.offsets.Day()
100 loops, best of 3: 5.72 ms per loop

In [3]: %timeit idx + pd.Timedelta(1, 'D')
100 loops, best of 3: 5.67 ms per loop

In [4]: %timeit idx + pd.offsets.DateOffset(years=1, months=2, days=1)
1 loops, best of 3: 1.79 s per loop

In [5]: %timeit idx + pd.offsets.WeekOfMonth(week=2, weekday=4)
1 loops, best of 3: 22.8 s per loop
```
### PR

```
In [3]: %timeit idx + pd.offsets.Day()
100 loops, best of 3: 5.8 ms per loop

In [4]: %timeit idx + pd.Timedelta(1, 'D')
100 loops, best of 3: 5.73 ms per loop

In [5]: %timeit idx + pd.offsets.DateOffset(years=1, months=2, days=1)
10 loops, best of 3: 86.9 ms per loop

In [6]: %timeit idx + pd.offsets.WeekOfMonth(week=2, weekday=4)
1 loops, best of 3: 23.4 s per loop
c:\users\chris\documents\python-dev\pandas\pandas\tseries\index.py:695: PerformanceWarning:     Non-vectorized DateOffset being applied to Series or DatetimeIndex  PerformanceWarning)
```

chris-b1 committed:
ENH: Allow DateOffset operations with Series

jreback commented:
This is a nice idea, but it wasn't originally implemented because it has massive perf issues.
However, a divide & conquer strategy can fix most of this. 

```
In [61]: np.random.seed(1234)

In [62]: s = Series(pd.to_timedelta(np.random.randint(0,10000000,size=10000),unit='s') + Timestamp('20130101'))

In [63]: pd.set_option('max_rows',10)

In [64]: np.random.seed(1234)

In [65]: s = Series(pd.to_timedelta(np.random.randint(0,10000000,size=10000),unit='s') + Timestamp('20130101'))

In [66]: s
Out[66]: 
0      2013-01-06 15:03:11
1      2013-03-20 00:58:59
2      2013-02-20 10:57:42
3      2013-03-13 04:40:21
4      2013-01-12 07:42:04
               ...        
9995   2013-02-17 07:22:39
9996   2013-04-03 04:57:11
9997   2013-04-16 10:06:38
9998   2013-02-15 15:28:46
9999   2013-03-05 12:00:39
dtype: datetime64[ns]
```

Any offset that has a ``_inc`` (e.g. <= Day), is simply convertible to a ``Timedelta``. (this *might* work for Week as well as its a fixed interval).

Note that your PR is equivalent to item by item apply.

```
In [78]: %timeit s.apply(lambda x: x+pd.offsets.Day())
1 loops, best of 3: 347 ms per loop

In [79]: %timeit s+Timedelta('1D')
1000 loops, best of 3: 610 us per loop
```

Conversions to > Day are a bit tricky. Note that you need to convert using ``pd.DatetimeIndex()`` because the obvious ``s.dt.to_period()`` round-trips thru ``object`` which defeats the purpose.
This would do a MonthEnd conversion. I think some of this could be make easier with some additonal methods (prob to ``DatetimeIndex``)

```
In [67]: i = pd.DatetimeIndex(s)

In [68]: base = (((i.to_period('M')+1).to_timestamp()).to_period('D')-1).to_timestamp()

In [69]: off = pd.TimedeltaIndex(i.asi8-i.to_period('D').to_timestamp().asi8)

In [70]: result = s.apply(lambda x: x+pd.offsets.MonthEnd())

In [71]: %timeit s.apply(lambda x: x+pd.offsets.MonthEnd())
1 loops, best of 3: 569 ms per loop
In [72]: result
Out[72]: 
0      2013-01-31 15:03:11
1      2013-03-31 00:58:59
2      2013-02-28 10:57:42
3      2013-03-31 04:40:21
4      2013-01-31 07:42:04
               ...        
9995   2013-02-28 07:22:39
9996   2013-04-30 04:57:11
9997   2013-04-30 10:06:38
9998   2013-02-28 15:28:46
9999   2013-03-31 12:00:39
dtype: datetime64[ns]

In [74]: Series(base+off)
Out[74]: 
0      2013-01-31 15:03:11
1      2013-03-31 00:58:59
2      2013-02-28 10:57:42
3      2013-03-31 04:40:21
4      2013-01-31 07:42:04
               ...        
9995   2013-02-28 07:22:39
9996   2013-04-30 04:57:11
9997   2013-04-30 10:06:38
9998   2013-02-28 15:28:46
9999   2013-03-31 12:00:39
dtype: datetime64[ns]

In [77]: %timeit Series((((i.to_period('M')+1).to_timestamp()).to_period('D')-1).to_timestamp() + pd.TimedeltaIndex(i.asi8-i.to_period('D').to_timestamp().asi8))
100 loops, best of 3: 7.74 ms per loop
```

Finally, you need to do Anchored offsets like ``BusinessDay``, ``W-THU`` via ``.apply`` (.e.g. I am not sure these can be vectorized, though maybe for something like ``W-THU`` you could do it).

You should emit a ``PerformanceWarning`` if you accept these last ones (or just raise).

jreback commented:
Furthermore this would need tests for each of the offsets as well as subtracting.

jreback commented:
this could be a property of a ``DateimeIndex``, it is equivalent of ``.time``, but returns a much more useful format ``pd.TimedeltaIndex(i.asi8-i.to_period('D').to_timestamp().asi8)``, maybe ``.timedelta`` (though not that descsriptive)

chris-b1 commented:
Thanks.  For those offsets that can be vectorized, where would that logic sit?  Something like a `vectorized_apply` method that `DateOffset.__add__` calls to if the other is array-like?

jreback commented:
you can prob put it in the same place (but maybe put in a separate method in ``_Timeop``). Though some of the functions can put put in ``DatetimeIndex`` i think.

chris-b1 commented:
@jreback can you take a look at the latest updates  It might not be all the way there, but implements most of the obvious fastpaths I could see.

Instead of `.timedelta` like you showed  I added a `period_delta(freq)` method to `DatetimeIndex` which is the same idea, but handles other frequencies too ('M' and 'W' used).  Wasn't sure if this should be public.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -133,6 +133,8 @@ Other enhancements
 
 - ``to_datetime`` can now accept ``yearfirst`` keyword (:issue:`7599`)
 
+- ``pandas.tseries.offset`` can now be used with with ``Series`` (:issue:`10699`)
```
say for add/sub ops

jreback commented on pandas/core/ops.py:
```diff
@@ -276,15 +276,35 @@ def __init__(self, left, right, name):
+
+        if self._is_offset(left):
+            lvalues = left
+            self.is_offset_lhs = True
```
maybe some of these up top to make this less repetive, e.g. (e.g. the 3 ``False`` ones)

```
self.is_datetime_lhs = False
....
```

jreback commented on pandas/tests/test_series.py:
```diff
@@ -3286,14 +3286,59 @@ def test_timedeltas_with_DateOffset(self):
+            else:
+                do = do
+                kwargs = {}
+            print do
```
remove the print

jreback commented on pandas/tseries/index.py:
```diff
@@ -681,6 +687,67 @@ def _add_delta(self, delta):
         return result
 
+    def _add_offset(self, offset):
+        from pandas import Timedelta
```
I would move this entire thing to the actual offsets themselves. E.g add a method in ``DateOffset``, maybe ``period_delta(DatetimeIndex i)``, e.g. it takes a ``DatetimeIndex`` (and for the ones that don't support, either raise a ``ValueError``. Add the ``_begend_offset`` (would split into 2 methods), add as an internal method of the offset as well.

jreback commented on pandas/tseries/index.py:
```diff
@@ -834,6 +901,22 @@ def union(self, other):
                     result.offset = to_offset(result.inferred_freq)
             return result
 
+    def period_delta(self, freq):
```
how about ``to_perioddelta``

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -307,6 +307,9 @@ def __call__(self, other):
 
     def __add__(self, other):
+        from pandas import Series
+        if isinstance(other, Series):
```
you can import ``core.common.ABCSeries`` at the top of the file and just check

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -2454,6 +2454,13 @@ def test_intersection_bug_1708(self):
+        from pandas import DateOffset
+        result = DatetimeIndex(['2000-01-01','2000-01-02']) + DateOffset(years=1)
+        exp = DatetimeIndex(['2001-01-01', '2001-01-02'])
+        self.assertTrue(result.equals(exp))
```
you can move all of the above tests (that you have in ``test_series``) here. Then loop thru both a ``Series`` and ``DatetimeIndex`` to do all of the ops. We do a lot of this in ``tseries/tests/test_base.py`` for example.

jreback commented on pandas/tseries/index.py:
```diff
@@ -681,6 +687,67 @@ def _add_delta(self, delta):
+                    if weeks:
+                        i = (i.to_period('W') + weeks).to_timestamp() + i.period_delta('W')
+
+                    d = (offset.kwds.get('days', 0)) * n
```
you can do this in the constructor of the ``Timedelta`` itself, e.g. something like

delta = ``Timedelta(**offset.kwds)``

jreback commented:
@chris-b1 looks pretty good. Just a bit or reorg to put the code in the appropriate spots.

jreback commented:
also if you'd add some vbenches would be great (and show the results in the very top of the PR).

chris-b1 commented:
@jreback  - see the update,  I think I've got most everything addressed, thanks for all the notes.

I wasn't sure what the `DateOffset` method that adds to a `DatetimeIndex` should be called - you suggested `period_delta` - I think that may be confusing where it doesn't return a `Timedelta`?  I ended up using `apply_index` but would be glad to change it to something better.

I'm on Windows and haven't able to get vbench working quite right - I did add some new benchmarks, and put timeits at the top of the PR.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -133,6 +133,8 @@ Other enhancements
 - ``to_datetime`` can now accept ``yearfirst`` keyword (:issue:`7599`)
 
+- ``pandas.tseries.offset`` can now be used with with ``Series`` for addition/subtraction (:issue:`10699`)
+
```
maybe clarify that it would work with <= Day previously, now works with M/Y

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -221,6 +223,39 @@ def apply(self, other):
             return other + timedelta(self.n)
 
+    def apply_index(self, i):
+        relativedelta_fast = set(['years', 'months', 'weeks',
```
can you add a doc-string

jreback replied:
where is ``relativedelta_fast`` actually used?

chris-b1 replied:
It's 4 lines down - if a keyword was passed to `DateOffset` that's valid in `relativedelta` but I can't handle (e.g. singular `day` argument) this is used to fallback.

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -363,6 +402,20 @@ def onOffset(self, dt):
+        base = (i.to_period(freq) + n).to_timestamp()
+        off = i.to_perioddelta('D')
+        return base + off
+    def _end_apply_index(self, i, freq):
```
blank line

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -363,6 +402,20 @@ def onOffset(self, dt):
 
+    # helpers for vectorized offset
+    def _beg_apply_index(self, i, freq):
+        n = self.n + (0 if self.n >= 0 else 1)
```
put a doc-string to explain each of these (could just be single line)

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -2454,6 +2454,61 @@ def test_intersection_bug_1708(self):
+                    do = do
+                    kwargs = {}
+                op = getattr(pd.offsets,do)(5, **kwargs)
+                self.assert_numpy_array_equal(s.map(lambda x: x + op), s + op)
```
these are going to show the ``PerformanceWarning`` yes?, wrap an ``assert_produces_warning(....)`` around whichever shows

jreback commented on pandas/core/ops.py:
```diff
@@ -276,15 +276,36 @@ def __init__(self, left, right, name):
+        self.is_integer_rhs = False
+
+        # DateOffsets skip array conversion
+        if self._is_offset(left):
```
why don't you add the variables: ``self.is_offset_rhs = self._is_offset(right)`` (and for the left), then just have ``self._convert_to_array(....)`` just return the data (w/o modification). I think a bit cleaner that way.

jreback commented:
looks pretty good. couple of code comments (mostly minor, you did nice job on moving things around!).
couple of doc additions.

chris-b1 commented:
@jreback  - here's another update - I addressed your comments, and expanded the tests, which highlighted a couple other changes needed.
  - I had broke adding offsets to `Timedelta`s
  - Handling normalized offsets
  - `month` kwarg in `YearEnd/Begin`; matching rollforwards when date is on `End/Begin` point

I also added vectorized paths for the `BusinessDay`, `Week`, `Quarter` offsets - same of approach of manipulating their `Period` equivalent.

jreback commented on pandas/core/ops.py:
```diff
@@ -332,6 +339,13 @@ def _validate(self):
+        elif ((self.is_datetime_lhs and self.is_offset_rhs) or
+              (self.is_datetime_rhs and self.is_offset_lhs)):
+            if self.name not in ('__add__', '__sub__'):
+                raise TypeError("can only operate on DateOffset for"
```
there is a section on datetime and timedelta where u can put this

chris-b1 replied:
moved

jreback commented on pandas/core/ops.py:
```diff
@@ -439,6 +461,18 @@ def f(x):
+            return True
+        elif isinstance(arr_or_obj, pd.Series):
+            return isinstance(arr_or_obj.values[0], pd.DateOffset)
+        elif is_list_like(arr_or_obj):
```
should check that all elements are offsets. there is an edge case where u have an array of date offsets (which may or may not be the same); I think in these cases u cannot use the vectorizes tricks
so what u are testing for here is a single offset

chris-b1 replied:
Changed this, added a test for `Series`.  Adding an array of offsets to a `DatetimeIndex` still fails, which I think is consistent with other `Index` behavior.

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -221,6 +232,60 @@ def apply(self, other):
+                                  'days', 'hours', 'minutes',
+                                  'seconds', 'microseconds'])
+        # relativedelta/_offset path only valid for base DateOffset
+        if type(self) is DateOffset:
```
use isinstance

chris-b1 replied:
The intent was to only execute this block for the base `DateOffset` class - it produces invalid deltas for subclasses that don't override `__init__` (e.g. `BusinessMonthBegin`).  Would it be better to explicitly raise `NotImplementedError` from those subclasses?

jreback replied:
yes I think those should raise ``NotImplementedError`` if they don't support a feature. You may have to an a method to ``DateOffset`` that raises for non-valid sub-classes (trying to think of a good word for those).

chris-b1 replied:
I apologize, I'm not completely clear - you're saying each subclass should implement its own `apply_index` method and raise `NotImplementedError`?  Or should that be handled some other way in the base `DateOffset` class?

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -221,6 +232,60 @@ def apply(self, other):
+                return i + (self._offset * self.n)
+            else:
+                # relativedelta with other keywords
+                raise NotImplementedError
```
can u provide a msg in the NotImplementedError

jreback commented on doc/source/timeseries.rst:
```diff
@@ -647,6 +647,33 @@ Another example is parameterizing ``YearEnd`` with the specific ending month:
+Using offsets with ``Series`` / ``DatetimeIndex``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Offsets can be used with either a ``Series`` or ``DatetimeIndex`` to
```
provide a link to this section as its pretty similar: http://pandas.pydata.org/pandas-docs/stable/timedeltas.html#operations (feel free to adjust language in either as needed)

jreback commented on pandas/core/ops.py:
```diff
@@ -407,13 +408,36 @@ def _convert_for_datetime(self, lvalues, rvalues):
+            # if adding single offset try vectorized path
+            # in DatetimeIndex; otherwise elementwise apply
+            if self.is_offset_lhs:
+                if len(lvalues) == 1:
```
in theory one could provide a series of offsets that are all the same (and equal to each other), but that is prob a special case.

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -221,6 +232,65 @@ def apply(self, other):
+                                     "does not have a vectorized "
+                                     "implementation"
+                                     % (self.__class__.__name__,))
+        else:
```
you don't need the else here.

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -2454,6 +2455,83 @@ def test_intersection_bug_1708(self):
+                                    pd.offsets.MonthEnd()])
+                exp = klass([Timestamp('2001-1-1'), Timestamp('2000-2-29')])
+                self.assert_numpy_array_equal(result, exp)
+
```
add an array of the *same* offset (just to show it works, in theory it could be done in a vectorized way, but as apply ok for now). also these will raise warnings, use ``assert_raises_warning(PerformanceWarning)`` where needed

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -2454,6 +2455,83 @@ def test_intersection_bug_1708(self):
+            result = s + pd.DateOffset(years=1)
+            result2 = pd.DateOffset(years=1) + s
+            exp = klass(date_range('2001-01-01', '2001-01-31'))
+            self.assert_numpy_array_equal(result, exp)
```
these should all use ``assert_series_equal`` (so construct the required Series), rather than using ``assert_numpy_array_equal``, which is not a strong enough guarantee

jreback commented:
@chris-b1 looking pretty good. couple of doc changes / test changes. Almost there!

chris-b1 commented:
@jreback thanks, latest has the changes to doc / testing from your comments.

jreback commented on pandas/tseries/index.py:
```diff
@@ -834,6 +848,22 @@ def union(self, other):
                     result.offset = to_offset(result.inferred_freq)
             return result
 
+    def to_perioddelta(self, freq):
```
can you add this to the API.rst, also add a versionadded

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -221,6 +232,65 @@ def apply(self, other):
+    def apply_index(self, i):
+        """
+        Vectorized apply of DateOffset to DatetimeIndex,
+        raises NotImplentedError for offsets without a
```
add versionadded

jreback commented:
small doc adds. pls squash. ping when green.

chris-b1 commented:
@jreback green - thanks!

jreback commented:
@chris-b1 great effort! sometimes these are involved issues!
